File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int low = 0, high = nums.length - 1, ans = Integer.MAX_VALUE;
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(nums[mid] < nums[high]){
                //Right Half is sorted & Nums[mid] would be smallest
                ans = Math.min(ans, nums[mid]);
                high = mid - 1;
            } else{
                // Left Half is sorted & Nums[low] would be smallest
                ans = Math.min(ans, nums[low]);
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size();
        int l = 0, r = n - 1, ans = INT_MAX;
        while(l <= r){
            int m = l + (r - l) / 2;
            if(nums[l] <= nums[m]){  //left sorted
                ans = min(ans, nums[l]); l = m + 1;
            } else{
                ans = min(ans, nums[m]); r = m - 1;
            }
        }
        return ans;
    }
};